codegenerator/cli/src/fuel/abi.rs

```diff
@@ -59,48 +60,26 @@ pub struct Abi {
     pub path_buf: PathBuf,
     pub path: String,
     pub raw: String,
-    program: ProgramABI,
+    program: UnifiedProgramABI,
     logs: HashMap<String, FuelLog>,
-    types: HashMap<usize, FuelType>,
+    types: HashMap<TypeId, FuelType>,
 }
 
 impl Abi {
-    fn decode_program(raw: &String) -> Result<ProgramABI> {
-        let program: ProgramABI = serde_json::from_str(&raw)?;
-        Ok(program)
+    fn decode_program(raw: &String) -> Result<UnifiedProgramABI> {
+        Ok(UnifiedProgramABI::from_json_abi(raw)?)
     }
 
-    fn decode_types(program: &ProgramABI) -> Result<HashMap<usize, FuelType>> {
-        let mut types_map: HashMap<usize, FuelType> = HashMap::new();
-
-        //eg "generic T" returns "T" for keyword "generic"
-        fn extract_value_after_keyword(keyword: &str, input: &str) -> Option<String> {
-            if let Some(start) = input.find(keyword) {
-                // Calculate the start position of the value after "keyword"
-                let value_start = start + keyword.len();
-                // Extract the value and trim any leading/trailing whitespace
-                let value = input[value_start..].trim();
-                if !value.is_empty() {
-                    return Some(value.to_string());
-                }
-            }
-            None
-        }
+    fn decode_types(program: &UnifiedProgramABI) -> Result<HashMap<TypeId, FuelType>> {
+        let mut types_map: HashMap<TypeId, FuelType> = HashMap::new();
 
-        fn mk_type_id_name(type_id: &usize) -> String {
-            format!("type_id_{}", type_id)
+        fn mk_type_id_name(type_id: &TypeId) -> String {
+            match type_id {
+                TypeId::Concrete(ConcreteTypeId(id)) => format!("concrete_type_id_{}", id),
+                TypeId::Metadata(MetadataTypeId(id)) => format!("metadata_type_id_{}", id),
```

Comment:
It would still be nice to make these meaningful names in future. I never got round to it on the first implementaion.